Implementation/Services/index.md

```diff
@@ -0,0 +1,510 @@
+                if (contentItem.ContentType.Alias == "landingPage")
+                {
+                    // we have injected in the mediaService in the contstructor for the component see above.
+                   bool hasExistingFolder = _mediaService.GetByLevel(1).Any(f => f.Name == contentItem.Name);
+                   if (!hasExistingFolder)
+                    {
+                        //let's create one (-1 indicates the root of the media section)
+                       IMedia newFolder = _mediaService.CreateMedia(contentItem.Name, -1, "Folder");
+                        _mediaService.Save(newFolder);
+                    }
+                }
+            }
+        }       
+
+        public void Terminate()
+        {
+           
+        }
+    }
+}
+```
+
+Management Services that are not transient (eg do not rely on UmbracoContext), eg `ContentService`, `MediaService` can be injected anywhere.
+
+### Accessing Published Content Cache from a Component or C# Class
+
+Injecting `UmbracoHelper` or `IPublishedContentQuery` into classes that aren't guaranteed to have an UmbracoContext will trigger a boot error, however there is a technique that allows the querying of the Umbraco Published Content Cache, using the `UmbracoContextFactory` and calling `EnsureUmbracoContext()`.
+
+```csharp
+using System.Collections.Generic;
+using Umbraco.Core.Composing;
+using Umbraco.Core.Models.PublishedContent;
+using Umbraco.Core.Services;
+using Umbraco.Core.Services.Implement;
+using Umbraco.Web;
+
+namespace Umbraco8.Components
+{
+    //adds the component to Umbraco composition's list of components
+    public class PublishEventsComposer : ComponentComposer<PublishEventsComponent>
+    {
+
+    }
+    //the component safely accessing the published content cache
+    public class PublishEventsComponent : IComponent
+    {
+        private readonly IUmbracoContextFactory _umbracoContextFactory;
+
+        public PublishEventsComponent(IUmbracoContextFactory umbracoContextFactory)
+        {
+            _umbracoContextFactory = umbracoContextFactory;
+        }
+
+        public void Initialize()
+        {
+            ContentService.Published += ContentService_Published;
+        }
+
+        private void ContentService_Published(IContentService sender, Umbraco.Core.Events.ContentPublishedEventArgs e)
+        {
+            //first call EnsureUmbracoContext
+            using (var umbracoContextReference = _umbracoContextFactory.EnsureUmbracoContext())
+            {
+                //the UmbracoContextReference provides access to the ContentCache
+                var contentCache = umbracoContextReference.UmbracoContext.ContentCache;
+                //query the published content cache
+                IEnumerable<IPublishedContent> rootNodes = contentCache.GetAtRoot();
+                //retrieve a particular item with id 1234 from the published content cache
```

Comment:
It isn't trying to suggest that, it is just a shit example! - just showing the technique without any real thought for the context of it's usage, other than somehere that is outside of a request - you would never really 'do that' kind of thing  in 'that event' - had a bit of a think - keen not to guide people to be sending emails 'on publish' and my CacheUpdated examples seemed too convoluted ... examples are hard too...

... what I came up with, which may not be desirable either ...

is when unpublishing event fires - you might want to retrieve what the published Url once was to the unpublished item to store in a db table to be able to serve proper 410 'page gone' status codes via an IContentFinder... for future requests instead of a 404. It's easily understandable as a concept but is getting the IPublishedContent object from content cache, in order to read the Url convoluted? or make sense in the more complex world of variants... 

just needs to be something not awful, but allows the reader to focus on the technique not what is trying to be done by the example...

I've put that into the PR anyway, it is better than what was there.